File: src/main/java/com/ethanmcdonough/ReportQueue/models/DataInserter.java
package com.ethanmcdonough.ReportQueue.models;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;
import java.util.concurrent.ExecutionException;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.json.Json;
import javax.json.JsonObject;
import javax.json.JsonReader;

import javax.servlet.http.HttpServletResponse;

import com.github.scribejava.core.model.OAuth1AccessToken;
import com.github.scribejava.core.model.OAuthRequest;
import com.github.scribejava.core.model.Verb;
import com.github.scribejava.core.oauth.OAuth10aService;
import com.github.scribejava.core.model.Response;

import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.util.store.FileDataStoreFactory;
import com.google.api.services.sheets.v4.SheetsScopes;
import com.google.api.services.sheets.v4.model.*;
import com.google.api.services.sheets.v4.Sheets;

public abstract class DataInserter {
	private File googleCredentials, googleClientSecret;
	private OAuth10aService kaservice;
	private static final List<String> googleScopes = Arrays.asList(SheetsScopes.SPREADSHEETS, SheetsScopes.DRIVE);
	private static final JsonFactory googleJsonFactory = JacksonFactory.getDefaultInstance();
	private static HttpTransport googleHttpTransport;
	private String appName, spreadsheetId;
	private static final SimpleDateFormat datef = new SimpleDateFormat("MM-dd-yyyy HH:mm:ss zzz");

	private static final int userReasonCharCount = 500;

	private static final long userReportTimeMillis = minToMillis(60);
	private static final int userReportTimeCount = 5;

	protected static final Logger log = Logger.getLogger("log");

	static {
		datef.setTimeZone(TimeZone.getTimeZone("GMT"));
		try {
			googleHttpTransport = GoogleNetHttpTransport.newTrustedTransport();
		} catch (Exception e) {
			log.log(Level.SEVERE, "Could not initialize googleHttpTransport", e);
		}
	}

	private static long minToMillis(int minutes) {
		return minutes * 60000L;
	}

	public DataInserter(String appName, String spreadsheetId, OAuth10aService kaservice, File googleClientSecret,
			File googleCredentials) {
		this.appName = appName;
		this.spreadsheetId = spreadsheetId;
		this.googleClientSecret = googleClientSecret;
		this.googleCredentials = googleCredentials;
		this.kaservice = kaservice;
	}

	public DataInserterResponse serve(OAuth1AccessToken accessToken, String id, String reason, String ip) {
		if (!validateId(id)) {
			return new DataInserterResponse(HttpServletResponse.SC_BAD_REQUEST, "Invalid id", false);
		}

		if (reason.length() > userReasonCharCount) {
			return new DataInserterResponse(HttpServletResponse.SC_BAD_REQUEST,
					"Your reason cannot exceed " + userReasonCharCount + " characters in length", false);
		}

		String insertId;
		try {
			insertId = getInsertId(id);
		} catch (Exception e) {
			return new DataInserterResponse(HttpServletResponse.SC_BAD_REQUEST, "Invalid id", false);
		}

		String kaid;
		try {
			kaid = getSenderKaid(accessToken);
		} catch (Exception e) {
			log.log(Level.WARNING, "Could not get user KAID from access token", e);
			return new DataInserterResponse(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized", false);
		}

		Sheets sheetService;
		try {
			sheetService = getSheetsService();
		} catch (IOException e) {
			log.log(Level.WARNING, "Could not get Google sheets service", e);
			return new DataInserterResponse(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Internal server error",
					false);
		}

		List<List<Object>> currentEntries;
		try {
			currentEntries = getCurrentEntries(sheetService);
		} catch (IOException e) {
			log.log(Level.WARNING, "Could not get existing entries", e);
			return new DataInserterResponse(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Internal server error",
					false);
		}

		Date dateInserted = new Date();

		if (testIfAlreadySubmitted(currentEntries, kaid, id)) {
			return new DataInserterResponse(HttpServletResponse.SC_BAD_REQUEST,
					String.format("You already reported that %s", getType()), false);
		}

		if (testIfQuotaExceeded(currentEntries, kaid, dateInserted)) {
			return new DataInserterResponse(HttpServletResponse.SC_BAD_REQUEST,
					String.format("You cannot report more than %s times per hour.", userReportTimeCount), false);
		}

		try {
			insertItems(sheetService, insertId, kaid, reason, ip, dateInserted);
		} catch (IOException e) {
			log.log(Level.WARNING, "Could not insert data", e);
			return new DataInserterResponse(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Internal server error",
					false);
		}

		return new DataInserterResponse(HttpServletResponse.SC_OK, "Thanks.  A Guardian will review your report.",
				true);
	}

	private String getSenderKaid(OAuth1AccessToken accessToken)
			throws InterruptedException, ExecutionException, IOException {
		OAuthRequest req = new OAuthRequest(Verb.GET, "https://www.khanacademy.org/api/v1/user");
		kaservice.signRequest(accessToken, req);
		Response res = kaservice.execute(req);
		InputStream resIS = res.getStream();
		JsonReader reader = Json.createReader(resIS);
		JsonObject responseJSON = reader.readObject();
		reader.close();
		resIS.close();
		return responseJSON.getString("kaid");
	}

	private Credential getCredentials() throws IOException {
		InputStream clientSecret = null;
		clientSecret = new FileInputStream(googleClientSecret);

		File credentials = googleCredentials;
		FileDataStoreFactory googleDataStoreFactory = new FileDataStoreFactory(credentials);
		GoogleClientSecrets googleClientSecrets = GoogleClientSecrets.load(googleJsonFactory,
				new InputStreamReader(clientSecret));

		clientSecret.close();

		GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(googleHttpTransport,
				googleJsonFactory, googleClientSecrets, googleScopes).setDataStoreFactory(googleDataStoreFactory)
						.setAccessType("offline").build();

		return new AuthorizationCodeInstalledApp(flow, new LocalServerReceiver()).authorize("user");
	}

	private Sheets getSheetsService() throws IOException {
		return new Sheets.Builder(googleHttpTransport, googleJsonFactory, getCredentials()).setApplicationName(appName)
				.build();
	}

	private List<List<Object>> getCurrentEntries(Sheets service) throws IOException {
		ValueRange currentData = service.spreadsheets().values().get(spreadsheetId, "A:E").execute();
		List<List<Object>> cdata = new ArrayList<List<Object>>(currentData.getValues());
		if (cdata.get(0) != null)
			cdata.remove(0);
		return cdata;
	}

	private void insertItems(Sheets service, Object id, Object reportingKaid, Object reason, Object ip, Date date)
			throws IOException {
		ValueRange insertionBody = new ValueRange()
				.setValues(Arrays.asList(Arrays.asList(id, reportingKaid, reason, ip, (Object) datef.format(date))));

		Sheets.Spreadsheets.Values.Append insertRequest = service.spreadsheets().values().append(spreadsheetId, "A:E",
				insertionBody);
		insertRequest.setValueInputOption("USER_ENTERED");
		insertRequest.setInsertDataOption("OVERWRITE");

		insertRequest.execute();
	}

	private boolean testIfQuotaExceeded(List<List<Object>> data, String kaid, Date date) {
		final String finKaid = kaid;
		final Date insertDate = date;
		return data.stream().filter(e -> {
			try {
				Date reportedDate = datef.parse((String) e.get(4));
				return e.get(1) != null && e.get(1).equals(finKaid) && e.get(4) != null && reportedDate != null
						&& insertDate.getTime() - reportedDate.getTime() < userReportTimeMillis;
			} catch (Exception e1) {
				log.log(Level.INFO, "Could not parse " + ((String) e.get(4)), e);
				return false;
			}
		}).count() >= userReportTimeCount;
	}

	private boolean testIfAlreadySubmitted(List<List<Object>> data, String kaid, String id) {
		final String finKaid = kaid, finId = id;
		return data.stream().filter(
				e -> isModifiedId(finId, (String) e.get(0)) && e.get(1) != null && ((String) e.get(1)).equals(finKaid))
				.count() > 0;
	}

	abstract protected boolean validateId(String id);

	abstract protected boolean isModifiedId(String id, String modifiedId);

	abstract protected String getType();

	abstract protected String getInsertId(String id) throws Exception;
}

File: server/app/models/DataInserter.scala
package models

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Arrays.asList
import java.io.{ IOException, FileInputStream, File, InputStreamReader }

import play.Logger

import ioutil.IOUtil.using

import scala.util.Try
import scala.collection.JavaConverters.{ asScalaBuffer, asJavaCollection }
import scala.collection.mutable.ListBuffer

import com.google.api.client.googleapis.auth.oauth2.{ GoogleAuthorizationCodeFlow, GoogleClientSecrets }
import com.google.api.services.sheets.v4.{ Sheets, SheetsScopes }
import com.google.api.client.json.JsonFactory
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.api.client.auth.oauth2.Credential
import com.google.api.client.util.store.FileDataStoreFactory
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver
import com.google.api.client.http.HttpTransport
import com.google.api.services.sheets.v4.model.ValueRange

class DataInserter(spreadsheetId: String, dataStore: File, clientSecret: File, httpTransport: HttpTransport, reporter: Reporter) {
  private val appName: String = "Report Queue"
  private val scopes: List[String] = List(SheetsScopes.SPREADSHEETS, SheetsScopes.DRIVE)
  private val jackson: JsonFactory = JacksonFactory.getDefaultInstance
  private val datef: SimpleDateFormat = new SimpleDateFormat("MM-dd-yyyy HH:mm:ss zzz")
  private val userReasonCharCount: Int = 500
  private val userReportTimeMillis: Long = 60 * 60000L
  private val userReportTimeCount: Int = 5

  @throws[IOException]
  def credentials: Credential = using[FileInputStream, Credential](new FileInputStream(clientSecret)) { clientSecret =>
    val googleDataStoreFactory: FileDataStoreFactory = new FileDataStoreFactory(dataStore)
    val googleClientSecrets: GoogleClientSecrets = GoogleClientSecrets.load(jackson, new InputStreamReader(clientSecret))
    val flow: GoogleAuthorizationCodeFlow = new GoogleAuthorizationCodeFlow.Builder(httpTransport, jackson, googleClientSecrets, asJavaCollection(scopes)).setDataStoreFactory(googleDataStoreFactory).setAccessType("offline").build
    new AuthorizationCodeInstalledApp(flow, new LocalServerReceiver).authorize("user")
  }

  @throws[IOException]
  def insert(service: Sheets, user: User, id: String, date: Date, reason: String, ip: String): Either[String, String] = {
    val data: List[List[Object]] = currentEntries(service)
    if (alreadySubmitted(data, user, id)) Left("You already reported that")
    else if (reason.length > userReasonCharCount) Left(s"Reasons can't exceed $userReasonCharCount characters in length")
    else if (quotaExceeded(data, user, date)) Left(s"You can't submit more than $userReportTimeCount reports in ${userReportTimeMillis / 60000L} minutes")
    else {
      reporter.transform(id) match {
        case Some(id) => {
          insertItems(service, id, user, reason, ip, date)
          Right("Thanks.  A Guardian will review your report.")
        }
        case None => Left("Invalid ID")
      }
    }
  }

  @throws[IOException]
  def service: Sheets = new Sheets.Builder(httpTransport, jackson, credentials).setApplicationName(appName).build

  @throws[IOException]
  def currentEntries(service: Sheets): List[List[Object]] = asScalaBuffer(service.spreadsheets.values.get(spreadsheetId, "A:E").execute.getValues).toList.map(asScalaBuffer(_).toList).drop(1)

  def quotaExceeded(data: List[List[Object]], user: User, date: Date): Boolean = data.count(e => Try(e(1).toString == user.kaid && date.getTime - datef.parse(e(4).toString).getTime < userReportTimeMillis) getOrElse {
    Logger.error("Failed to process row: " + e)
    false
  }) >= userReportTimeCount

  def alreadySubmitted(data: List[List[Object]], user: User, id: String): Boolean = data.count(e => (for {
    sheetReportedId <- (e lift 0).map(e => reporter.extractFromTransformed(e.toString)).flatten
    userKaid <- (e lift 1)
  } yield sheetReportedId == id && userKaid.toString == user.kaid) getOrElse false) > 0

  def insertItems(service: Sheets, id: String, user: User, reason: String, ip: String, date: Date): Unit = {
    val insertionBody: ValueRange = (new ValueRange).setValues(asList(asList(id.asInstanceOf[Object], user.kaid.asInstanceOf[Object], reason.asInstanceOf[Object], ip.asInstanceOf[Object], datef.format(date).asInstanceOf[Object])))
    val insertRequest = service.spreadsheets.values.append(spreadsheetId, "A:E", insertionBody)
    insertRequest.setValueInputOption("USER_ENTERED")
    insertRequest.setInsertDataOption("OVERWRITE")
    insertRequest.execute
  }
}

object DataInserter {
  def apply(spreadsheetId: String, dataStore: File, clientSecret: File, httpTransport: HttpTransport, reporter: Reporter): DataInserter = new DataInserter(spreadsheetId, dataStore, clientSecret, httpTransport, reporter)
}